keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -143,29 +144,51 @@ The CRD `v1` API will be the same as the `v1beta1` but with all changes to the A
   * Ensure discovery docs track a CRD through creation, version addition, version removal, and deletion
   * Ensure `spec.served` is respected
+* Tests for malformed inputs and errors (system should return sensible errors, not panic, and recover appropriately)
+  * TODO
 
 ### e2e tests for GA
 
-* Custom Resources should be readable and writable at all available versions (test for round trip-ability)
 * A Custom Resource watch should terminate if its CustomResourceDefinition is deleted or updated
 
 ### Scale Targets for GA
 
-* TODO quantify: Read/write latency of CRDs within X% of native Kubernetes types
-* TODO quantify: Latency degrades less than X% for up to 100k Custom Resources per CRD kind
-* TODO quantify: Webhook conversion QPS of a noop converter is within X% of QPS with no webhook
-* Coordinate with sig-scalability
+The scale target for GA of custom resources are defined by the same [API call latency
+SLIs/SLOs as the Kuberetes native types](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md#api-call-latency-slisslos-details):
+
+| SLI | SLO |
+| --- | --- |
+| Latency<sup>[1](#footnote1)</sup> of mutating<sup>[2](#footnote2)</sup> API calls for single objects for every (resource, verb) pair, measured as 99th percentile over last 5 minutes | In default Kubernetes installation, for every (resource, verb) pair, excluding virtual and aggregated resources and Custom Resource Definitions, 99th percentile per cluster-day <= 1s |
+| Latency<sup>[1](#footnote1)</sup> of non-streaming read-only<sup>[3](#footnote3)</sup> API calls for every (resource, scope<sup>[4](#footnote4)</sup>) pair, measured as 99th percentile over last 5 minutes | In default Kubernetes installation, for every (resource, scope) pair, excluding virtual and aggregated resources and Custom Resource Definitions, 99th percentile per cluster-day (a) <= 1s if `scope=resource` (b) <= 5s if `scope=namespace` (c) <= 30s if `scope=cluster` |
+
+And with these thresholds:
+
+| Quantity | Threshold scope=namespace | Threshold: scope=cluster |
+| --- | --- | --- |
+| Custom Resource objects per Definition | 1000 | 5000 |
+| Custom Resource Definitions | 100 | 500 |
+
+(See the [Kubernetes native type thresholds](https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md#kubernetes-thresholds) for comparison),
+
+Given that (a) the performance and scalability of conversion webhooks are the
+responsibility of their author, and (b) custom resources can be arbitrarily large, CRD scale targets that apply under these conditions:
+
+- custom resource object size <= 50kb
+- conversion webhook latency includes serialization/deserialization cost, but does not include the webhook conversion operation cost (i.e. the cost of the custom authored conversion routines for a particular conversion webhook is not accounted for)
```

Comment:
we should explicitly say excluding the latency caused by rate-limiter of the webhook client? iiuc, it's applying default QPS/Burst (5.0/10), will probably be a factor slowing down CR operations

Reply:
I'll add some fine print along the lines of "These SLIs are "server-side". Client latency might be higher to network, serialization costs, client side rate limiting and such.".

Reply:
yes we had to change the QPS and burst to 10k and 20k in our tests :)

Reply:
> I'll add some fine print along the lines of "These SLIs are "server-side". Client latency might be higher to network, serialization costs, client side rate limiting and such.".

That's not really true - the way we measure latency of api calls includes the Write() call, which actually is serializing data and sending them to client (and waiting for tcp ACK for those).
[I have an AI on myself to rethink it a bit more holistically too.]

Reply:
Oh right, I wasn’t thinking things through. Our tests do test from a client incl deserialization. We ran it on the master VM to minimize network noise.  I’ll update.